{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Joshua Burkhart  \n",
    "BMI 650  \n",
    "11/18/2015  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Midterm 2015"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1: Determine length of shortest subsequence that occurs exactly once in a genomic sequence of length N. Provide assumptions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['WR', 'RX', 'BW']\n",
      "2\n"
     ]
    }
   ],
   "source": [
    "s = 'ABABABABABABABWWWWWWWWRRRRRRRXXXXXXXX'\n",
    "\n",
    "def short_u_sub_strs(s):\n",
    "    \"\"\"\n",
    "    This function assumes |s| is finite and fits in RAM.\n",
    "    \n",
    "    Runtime is bound by the triangle number of |s|,\n",
    "    approx. O((n^2)/2), though an average case execution\n",
    "    would run in O(((n/2)^2)/2) = O((n^2)/8).\n",
    "    :param s: input string\n",
    "    :return: list of shortest unique substrings\n",
    "    \"\"\"\n",
    "    u_sub_strs = []\n",
    "    for l in range(1,len(s)+1):\n",
    "        found_u_sub_strs = False\n",
    "        sub_strs = {}\n",
    "        for i in range((len(s)-l)+1):\n",
    "            sub_str = str(s[i:i+l])\n",
    "            if sub_str in sub_strs:\n",
    "                sub_strs[sub_str] += 1\n",
    "            else:\n",
    "                sub_strs[sub_str] = 1\n",
    "        for sub_str in sub_strs:\n",
    "            if sub_strs[sub_str] == 1:\n",
    "                u_sub_strs.append(sub_str)\n",
    "                found_u_sub_strs = True\n",
    "        if found_u_sub_strs:\n",
    "            break\n",
    "    return u_sub_strs\n",
    "\n",
    "shortlist = short_u_sub_strs(s)\n",
    "\n",
    "print(shortlist)\n",
    "print(len(shortlist[0]) if len(shortlist) > 0 else 0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2: Regarding HG38 and CD8B"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### a. What chromosome is this gene on?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The below link shows it's on chromosome 2.\n",
    "\n",
    "https://genome.ucsc.edu/cgi-bin/hgc?hgsid=454559201_hi3ZA8G9DUTVIRE16jb1ZW8vB2jf&c=chr2&o=86815336&t=86861924&g=refGene&i=NM_172213"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### b. What is the start and end coordinate of this gene?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Transcript links (bottom) show positions:\n",
    "\n",
    "chr2:86815337-86861924\n",
    "chr2:86815337-86861924\n",
    "chr2:86815337-86861924\n",
    "chr2:86815337-86861924\n",
    "chr2:86841566-86861924\n",
    "\n",
    "The widest range of these coordinates is 86815337-86861924. As all transcripts are shown to be on the negative strand, we'll report\n",
    "the start coordinate as 86861924 and the end as 86815337.\n",
    "\n",
    "https://genome.ucsc.edu/cgi-bin/hgc?hgsid=454559201_hi3ZA8G9DUTVIRE16jb1ZW8vB2jf&c=chr2&o=86815336&t=86861924&g=refGene&i=NM_172213\n",
    "https://genome.ucsc.edu/cgi-bin/hgc?hgsid=454559201_hi3ZA8G9DUTVIRE16jb1ZW8vB2jf&c=chr2&o=86815336&t=86861924&g=refGene&i=NM_172102\n",
    "https://genome.ucsc.edu/cgi-bin/hgc?hgsid=454559201_hi3ZA8G9DUTVIRE16jb1ZW8vB2jf&c=chr2&o=86815336&t=86861924&g=refGene&i=NM_172101\n",
    "https://genome.ucsc.edu/cgi-bin/hgc?hgsid=454559201_hi3ZA8G9DUTVIRE16jb1ZW8vB2jf&c=chr2&o=86815336&t=86861924&g=refGene&i=NM_001178100\n",
    "https://genome.ucsc.edu/cgi-bin/hgc?hgsid=454559201_hi3ZA8G9DUTVIRE16jb1ZW8vB2jf&c=chr2&o=86841565&t=86861924&g=refGene&i=NM_004931"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### c. What strand of the DNA contains this gene?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Negative (see above links)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### d. How many transcripts are in this gene as reported by the RefSeq Consortium?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "5 (see above links)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### e. How many distinct exons are reported in the RefSeq transcripts?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Visualy inspecting the 5 transcripts using the UCSC genome browser (link at bottom) indicates  the first has 6 exons, the second has 5 exons (all shared with the first), the third has 7 exons (1 distinct and 6 shared with the first),  the fourth has 4 exons (all shared with the first), and the fifth has 6 exons (1 distinct and 5 shared with the first). Together, the transcripts have 8 distinct exons. \n",
    "\n",
    "https://genome.ucsc.edu/cgi-bin/hgTracks?db=hg38&position=chr2%3A86812162-86863658&hgsid=454559201_hi3ZA8G9DUTVIRE16jb1ZW8vB2jf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### f. How many distinct introns are reported in the RefSeq transcripts?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Again, visually inspecting the 5 transcripts using the UCSC genome browser (see above) shows the first has 5 introns, the second has 4 introns (1 distinct and 3 shared with the first), the third has 6 introns (2 distinct and 4 shared with the first), the fourth has 3 introns (1 distinct and 2 shared with the first), and the fifth has 5 introns (1 distinct and 4 shared with the first). Together, the transcripts have 10 distinct introns.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### g. List the genomic coordinates of each intron (start-end) from above which contains a canonical splice site."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Visually inspecting the transcript alignments (links at bottom) shows 9 introns are cononical, begining with GT and ending with AG. The coordinates are as follows:\n",
    "\n",
    "86861822 - 86858417 [gt-ag]  \n",
    "86858056 - 86853087 [gt-ag]  \n",
    "86852996 - 86846774 [gt-ag]  \n",
    "86846683 - 86844959 [gt-ag]  \n",
    "86844821 - 86815719 [gt-ag]  \n",
    "86852996 - 86844959 [gt-ag]  \n",
    "86844821 - 86822375 [gt-ag]  \n",
    "86852996 - 86815719 [gt-ag]  \n",
    "86844821 - 86842320 [gt-ag]  \n",
    "\n",
    "https://genome.ucsc.edu/cgi-bin/hgc?hgsid=454559201_hi3ZA8G9DUTVIRE16jb1ZW8vB2jf&g=htcCdnaAli&i=NM_172213&c=chr2&l=86800257&r=86903247&o=86815336&aliTable=refSeqAli&table=refGene\n",
    "https://genome.ucsc.edu/cgi-bin/hgc?hgsid=454559201_hi3ZA8G9DUTVIRE16jb1ZW8vB2jf&g=htcCdnaAli&i=NM_172102&c=chr2&l=86800257&r=86903247&o=86815336&aliTable=refSeqAli&table=refGene\n",
    "https://genome.ucsc.edu/cgi-bin/hgc?hgsid=454559201_hi3ZA8G9DUTVIRE16jb1ZW8vB2jf&g=htcCdnaAli&i=NM_172101&c=chr2&l=86800257&r=86903247&o=86815336&aliTable=refSeqAli&table=refGene\n",
    "https://genome.ucsc.edu/cgi-bin/hgc?hgsid=454559201_hi3ZA8G9DUTVIRE16jb1ZW8vB2jf&g=htcCdnaAli&i=NM_001178100&c=chr2&l=86800257&r=86903247&o=86815336&aliTable=refSeqAli&table=refGene\n",
    "https://genome.ucsc.edu/cgi-bin/hgc?hgsid=454559201_hi3ZA8G9DUTVIRE16jb1ZW8vB2jf&g=htcCdnaAli&i=NM_004931&c=chr2&l=86800257&r=86903247&o=86841565&aliTable=refSeqAli&table=refGene"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### h. List the genomic coordinates of each intron (start-end) from above which contains a non-canonical splice site."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "86822320 - 86844959 [AT...AGgt-agGC...AA]  \n",
    "\n",
    "(see links above)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 1) What is the official gene symbol of the nearest gene to CD8B?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Visually inspection using the UCSC genome browser (link below) shows several genes are nearby.\n",
    "\n",
    "Assuming the UCSC genome browser uses official gene symbols, ANAPC1P1 is the nearest in bp, though it's on the opposite (positive) strand.\n",
    "\n",
    "https://genome.ucsc.edu/cgi-bin/hgTracks?db=hg38&position=chr2%3A86753188-86926991&hgsid=454559201_hi3ZA8G9DUTVIRE16jb1ZW8vB2jf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 2) What strand of the DNA contains this nearest neighboring gene?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Positive (see above)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 3) How would the location of these genes be described in relation to one another?\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The first exons of CD8B and ANAPC1P1 are derived from the same area of the genome, though on different strands. Both CD8B and ANAPC1P1 extend in opposite directions on their respective strands.  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3: Describe minimum information needed to unambiguously define the location of a gene. Provide example.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "[Genome Assembly, Start Coordinate, Stop Coordinate]\n",
    "\n",
    "Example:\n",
    "\n",
    "Genome Assembly: GRCh38/hg38  \n",
    "Start Coordinate: chr2:86861924  \n",
    "Stop Coordinate: chr2:86815337  \n",
    "\n",
    "We can infer strand by testing if the start coordinate is in front of or behind the stop coordinate.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4: Write program to randomly generate two sequences which will result in..."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### a. 1) a better local alignment as compared to global alignment (using blosum62 scoring matrix), both 10 amino acids long."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "r: BMPTSQCVNS\n",
      "l: vvtTSQalxc\n"
     ]
    }
   ],
   "source": [
    "from random import randint\n",
    "\n",
    "SIGMA = 'ARNDCQEGHILKMFPSTWYVBZX'\n",
    "\n",
    "def r_char():\n",
    "    \"\"\"\n",
    "    Generates a random character\n",
    "    :return: a random character from SIGMA\n",
    "    \"\"\"\n",
    "    return SIGMA[randint(0,len(SIGMA)-1)]\n",
    "\n",
    "\n",
    "def random_sequence(l):\n",
    "    \"\"\"\n",
    "    Generates a random sequence\n",
    "    :param l: length of random sequence to build\n",
    "    :return: random sequence of length l\n",
    "    \"\"\"\n",
    "    r_seq = ''\n",
    "    for i in range(l):\n",
    "        r_seq += r_char()\n",
    "    return r_seq\n",
    "\n",
    "def local_align_sequence(r_seq):\n",
    "    \"\"\"\n",
    "    Matches middle third of the characters in r_seq\n",
    "    :param r_seq: any random sequence\n",
    "    :return: a sequence that will align locally to r_seq\n",
    "    \"\"\"\n",
    "    l_seq = ''\n",
    "    t3 = len(r_seq)\n",
    "    t1 = int(len(r_seq)/3)\n",
    "    t2 = 2*t1\n",
    "    for i in range(t1):\n",
    "        l_seq += r_char().lower()\n",
    "    for i in range(t1,t2):\n",
    "        l_seq += r_seq[i]\n",
    "    for i in range(t2,t3):\n",
    "        l_seq += r_char().lower()\n",
    "    return l_seq\n",
    "\n",
    "r = random_sequence(10)\n",
    "l = local_align_sequence(r)\n",
    "\n",
    "print('r: ' + r)\n",
    "print('l: ' + l)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### a. 2) Discuss criteria and any limitations in generating your sequences. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "A pseudorandom number generator (PRNG) is used to select all characters for an initial 'random' sequence. The PRNG is reused to select the first and last thirds of a second sequence of equal length, while the middle third is selected by simply using the characters from the first."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### a. 3) Align the two sequences using dynamic programming using the blosum62 scoring matrix. Show the DP matrix and the resultant alignment."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DP matrix:\n",
      "  *   B   M   P   T   S   Q   C   V   N   S \n",
      "* 0   -4  -8 -12 -16 -20 -24 -28 -32 -36 -40 \n",
      "V -4  0   0   0   0   0   0   0   0   0   0  \n",
      "V -8  0  1.0  0  0.0  0   0   0  4.0 0.0  0  \n",
      "T-12  0   0  0.0 5.0 1.0  0   0  0.0 4.0 1.0 \n",
      "T-16  0   0   0  5.0 6.0 2.0  0  0.0 0.0 5.0 \n",
      "S-20  0   0   0  1.0 9.0 6.0 2.0  0  1.0 4.0 \n",
      "Q-24  0  0.0  0   0  5.0 14.010.06.0 2.0 1.0 \n",
      "A-28  0   0   0  0.0 1.0 10.014.010.06.0 3.0 \n",
      "L-32  0  2.0  0   0   0  6.0 10.015.011.07.0 \n",
      "X-36  0   0  0.0 0.0 0.0 2.0 6.0 11.014.011.0\n",
      "C-40  0   0   0   0   0   0  11.07.0 10.013.0\n",
      "BMPTSQCVNS\n",
      "   TSQALXC\n",
      "Sum of Pairs = 13.0\n",
      "DP matrix:\n",
      "  *   B   M   P   T   S   Q   C   V   N   S \n",
      "* 0   -4  -8 -12 -16 -20 -24 -28 -32 -36 -40 \n",
      "V -4 -3.0-3.0-7.0-11.0-15.0-19.0-23.0-24.0-28.0-32.0\n",
      "V -8 -7.0-2.0-5.0-7.0-11.0-15.0-19.0-19.0-23.0-27.0\n",
      "T-12 -9.0-6.0-3.00.0 -4.0-8.0-12.0-16.0-19.0-22.0\n",
      "T-16 -13.0-10.0-7.02.0 1.0 -3.0-7.0-11.0-15.0-18.0\n",
      "S-20 -16.0-14.0-11.0-2.06.0 2.0 -2.0-6.0-10.0-11.0\n",
      "Q-24 -20.0-16.0-15.0-6.02.0 11.07.0 3.0 -1.0-5.0\n",
      "A-28 -24.0-20.0-17.0-10.0-2.07.0 11.07.0 3.0 0.0 \n",
      "L-32 -28.0-22.0-21.0-14.0-6.03.0 7.0 12.08.0 4.0 \n",
      "X-36 -32.0-26.0-24.0-18.0-10.0-1.03.0 8.0 11.08.0 \n",
      "C-40 -36.0-30.0-28.0-22.0-14.0-5.08.0 4.0 7.0 10.0\n",
      "BMPTSQCVNS\n",
      "VVTTSQALXC\n",
      "Sum of Pairs = 10.0\n"
     ]
    }
   ],
   "source": [
    "import math\n",
    "from random import randint\n",
    "\n",
    "SIGMA = 'ARNDCQEGHILKMFPSTWYVBZX'\n",
    "ILLEGAL_CHARS = ' -\\n'\n",
    "GAP_PENALTY = -4 # taken from blosum62 matrix\n",
    "BLOSUM62 = {frozenset({'R', 'P'}): -2.0, frozenset({'S', 'G'}): 0.0, frozenset({'T', 'K'}): -1.0,\n",
    "            frozenset({'N', 'G'}): 0.0, frozenset({'Z', 'G'}): -2.0, frozenset({'P', 'B'}): -2.0,\n",
    "            frozenset({'T', 'W'}): -2.0, frozenset({'N', 'D'}): 1.0, frozenset({'*', 'N'}): -4.0, frozenset({'L'}): 4.0,\n",
    "            frozenset({'Q', 'A'}): -1.0, frozenset({'N', 'K'}): 0.0, frozenset({'*', 'M'}): -4.0,\n",
    "            frozenset({'N', 'A'}): -2.0, frozenset({'Q', 'P'}): -1.0, frozenset({'X', 'A'}): 0.0,\n",
    "            frozenset({'Z', 'K'}): 1.0, frozenset({'*', 'K'}): -4.0, frozenset({'H', 'D'}): -1.0,\n",
    "            frozenset({'C', 'D'}): -3.0, frozenset({'B', 'X'}): -1.0, frozenset({'Q', 'R'}): 1.0,\n",
    "            frozenset({'W', 'P'}): -4.0, frozenset({'S', 'Y'}): -2.0, frozenset({'*', 'W'}): -4.0,\n",
    "            frozenset({'W', 'D'}): -4.0, frozenset({'K', 'M'}): -1.0, frozenset({'I', 'V'}): 3.0,\n",
    "            frozenset({'E', 'K'}): 1.0, frozenset({'C', 'Y'}): -2.0, frozenset({'R', 'D'}): -2.0,\n",
    "            frozenset({'Z', 'B'}): 1.0, frozenset({'P', 'G'}): -2.0, frozenset({'Q', 'C'}): -3.0,\n",
    "            frozenset({'R', 'S'}): -1.0, frozenset({'B', 'L'}): -4.0, frozenset({'R', 'Z'}): 0.0,\n",
    "            frozenset({'I', 'F'}): 0.0, frozenset({'S', 'A'}): 1.0, frozenset({'I', 'B'}): -3.0,\n",
    "            frozenset({'F', 'Y'}): 3.0, frozenset({'D', 'A'}): -2.0, frozenset({'E', 'C'}): -4.0,\n",
    "            frozenset({'H', 'E'}): 0.0, frozenset({'I', 'C'}): -1.0, frozenset({'H', 'I'}): -3.0,\n",
    "            frozenset({'Q', 'T'}): -1.0, frozenset({'V', 'D'}): -3.0, frozenset({'K', 'D'}): -1.0,\n",
    "            frozenset({'Q', 'V'}): -2.0, frozenset({'C', 'B'}): -3.0, frozenset({'*', 'V'}): -4.0,\n",
    "            frozenset({'X', 'Y'}): -1.0, frozenset({'F', 'V'}): -1.0, frozenset({'I', 'M'}): 1.0,\n",
    "            frozenset({'P', 'M'}): -2.0, frozenset({'E', 'W'}): -3.0, frozenset({'W', 'A'}): -3.0,\n",
    "            frozenset({'E', 'L'}): -3.0, frozenset({'T', 'Z'}): -1.0, frozenset({'*', 'E'}): -4.0,\n",
    "            frozenset({'I', 'N'}): -3.0, frozenset({'E', 'Z'}): 4.0, frozenset({'H', 'A'}): -2.0,\n",
    "            frozenset({'B', 'K'}): 0.0, frozenset({'M'}): 5.0, frozenset({'T', 'Y'}): -2.0, frozenset({'D', 'G'}): -1.0,\n",
    "            frozenset({'K'}): 5.0, frozenset({'W', 'N'}): -4.0, frozenset({'I', 'X'}): -1.0, frozenset({'A'}): 4.0,\n",
    "            frozenset({'N', 'L'}): -3.0, frozenset({'W', 'C'}): -2.0, frozenset({'H', 'W'}): -2.0,\n",
    "            frozenset({'F', 'M'}): 0.0, frozenset({'S', 'D'}): 0.0, frozenset({'*', 'D'}): -4.0,\n",
    "            frozenset({'P', 'X'}): -2.0, frozenset({'T', 'L'}): -1.0, frozenset({'L', 'M'}): 2.0,\n",
    "            frozenset({'D', 'M'}): -3.0, frozenset({'*', 'T'}): -4.0, frozenset({'P', 'Y'}): -3.0,\n",
    "            frozenset({'T', 'G'}): -2.0, frozenset({'R', 'Y'}): -2.0, frozenset({'I', 'P'}): -3.0,\n",
    "            frozenset({'*', 'S'}): -4.0, frozenset({'S', 'W'}): -3.0, frozenset({'F', 'G'}): -3.0,\n",
    "            frozenset({'I', 'T'}): -1.0, frozenset({'T', 'E'}): -1.0, frozenset({'G'}): 6.0, frozenset({'H', 'N'}): 1.0,\n",
    "            frozenset({'W', 'Y'}): 2.0, frozenset({'V', 'M'}): 1.0, frozenset({'N', 'C'}): -3.0,\n",
    "            frozenset({'V', 'L'}): 1.0, frozenset({'Q', 'Z'}): 3.0, frozenset({'R', 'G'}): -2.0,\n",
    "            frozenset({'F', 'X'}): -1.0, frozenset({'W', 'B'}): -4.0, frozenset({'H', 'B'}): 0.0,\n",
    "            frozenset({'V', 'A'}): 0.0, frozenset({'Z', 'Y'}): -2.0, frozenset({'W', 'L'}): -2.0,\n",
    "            frozenset({'K', 'A'}): -1.0, frozenset({'T', 'A'}): 0.0, frozenset({'Y', 'D'}): -3.0,\n",
    "            frozenset({'R', 'C'}): -3.0, frozenset({'X', 'M'}): -1.0, frozenset({'Q', 'L'}): -2.0,\n",
    "            frozenset({'Q', 'D'}): 0.0, frozenset({'I', 'W'}): -3.0, frozenset({'E', 'D'}): 2.0,\n",
    "            frozenset({'R', 'T'}): -1.0, frozenset({'Y', 'G'}): -3.0, frozenset({'S', 'E'}): 0.0,\n",
    "            frozenset({'F', 'P'}): -4.0, frozenset({'S'}): 4.0, frozenset({'W'}): 11.0, frozenset({'C', 'A'}): 0.0,\n",
    "            frozenset({'T', 'B'}): -1.0, frozenset({'R', 'F'}): -3.0, frozenset({'H', 'K'}): -1.0,\n",
    "            frozenset({'H', 'M'}): -2.0, frozenset({'C', 'M'}): -1.0, frozenset({'S', 'M'}): -1.0,\n",
    "            frozenset({'Q', 'I'}): -3.0, frozenset({'R'}): 5.0, frozenset({'F', 'C'}): -2.0,\n",
    "            frozenset({'H', 'F'}): -1.0, frozenset({'V', 'G'}): -3.0, frozenset({'Z', 'L'}): -3.0,\n",
    "            frozenset({'W', 'V'}): -3.0, frozenset({'*', 'Z'}): -4.0, frozenset({'H', 'C'}): -3.0,\n",
    "            frozenset({'N', 'V'}): -3.0, frozenset({'L', 'D'}): -4.0, frozenset({'R', 'V'}): -3.0,\n",
    "            frozenset({'Q', 'Y'}): -1.0, frozenset({'S', 'L'}): -2.0, frozenset({'N', 'M'}): -2.0,\n",
    "            frozenset({'Z', 'M'}): -1.0, frozenset({'F', 'L'}): 0.0, frozenset({'H', 'G'}): -2.0,\n",
    "            frozenset({'C', 'G'}): -3.0, frozenset({'F', 'Z'}): -3.0, frozenset({'V', 'B'}): -3.0,\n",
    "            frozenset({'H', 'L'}): -3.0, frozenset({'E'}): 5.0, frozenset({'*', 'C'}): -4.0,\n",
    "            frozenset({'H', '*'}): -4.0, frozenset({'F', 'B'}): -3.0, frozenset({'E', 'A'}): -1.0,\n",
    "            frozenset({'R', 'A'}): -1.0, frozenset({'H', 'R'}): 0.0, frozenset({'E', 'N'}): 0.0,\n",
    "            frozenset({'T', 'V'}): 0.0, frozenset({'*', 'B'}): -4.0, frozenset({'G', 'A'}): 0.0,\n",
    "            frozenset({'H', 'S'}): -1.0, frozenset({'Q', 'F'}): -3.0, frozenset({'N', 'P'}): -2.0,\n",
    "            frozenset({'D'}): 6.0, frozenset({'Z', 'X'}): -1.0, frozenset({'F', 'W'}): 1.0, frozenset({'N', 'Y'}): -2.0,\n",
    "            frozenset({'X', 'D'}): -1.0, frozenset({'Z', 'V'}): -2.0, frozenset({'Q', '*'}): -4.0,\n",
    "            frozenset({'Q', 'S'}): 0.0, frozenset({'L', 'K'}): -2.0, frozenset({'W', 'Z'}): -3.0,\n",
    "            frozenset({'*', 'Y'}): -4.0, frozenset({'*', 'L'}): -4.0, frozenset({'Q'}): 5.0,\n",
    "            frozenset({'L', 'A'}): -1.0, frozenset({'E', 'V'}): -2.0, frozenset({'I', 'Z'}): -3.0,\n",
    "            frozenset({'I', 'K'}): -3.0, frozenset({'E', 'Y'}): -2.0, frozenset({'C', 'K'}): -3.0,\n",
    "            frozenset({'I', 'Y'}): -1.0, frozenset({'*', 'G'}): -4.0, frozenset({'E', 'X'}): -1.0,\n",
    "            frozenset({'S', 'X'}): 0.0, frozenset({'Q', 'X'}): -1.0, frozenset({'P', 'A'}): -1.0,\n",
    "            frozenset({'F', 'K'}): -3.0, frozenset({'S', 'Z'}): 0.0, frozenset({'Z'}): 4.0, frozenset({'I', 'D'}): -3.0,\n",
    "            frozenset({'Q', 'H'}): 0.0, frozenset({'*', 'F'}): -4.0, frozenset({'F', 'D'}): -3.0,\n",
    "            frozenset({'T', 'D'}): -1.0, frozenset({'S', 'K'}): 0.0, frozenset({'I', '*'}): -4.0,\n",
    "            frozenset({'S', 'T'}): 1.0, frozenset({'C', 'L'}): -1.0, frozenset({'E', 'P'}): -1.0,\n",
    "            frozenset({'X'}): -1.0, frozenset({'R', 'X'}): -1.0, frozenset({'B', 'Y'}): -3.0,\n",
    "            frozenset({'R', 'L'}): -2.0, frozenset({'E', 'M'}): -2.0, frozenset({'*', 'R'}): -4.0,\n",
    "            frozenset({'H', 'V'}): -3.0, frozenset({'Q', 'W'}): -2.0, frozenset({'E', 'F'}): -3.0,\n",
    "            frozenset({'Z', 'D'}): 1.0, frozenset({'P', 'V'}): -2.0, frozenset({'V', 'K'}): -2.0, frozenset({'N'}): 6.0,\n",
    "            frozenset({'R', 'W'}): -3.0, frozenset({'S', 'P'}): -1.0, frozenset({'Z', 'A'}): -1.0,\n",
    "            frozenset({'P', 'D'}): -1.0, frozenset({'*', 'X'}): -4.0, frozenset({'V', 'Y'}): -1.0,\n",
    "            frozenset({'X', 'L'}): -1.0, frozenset({'I', 'G'}): -4.0, frozenset({'E', 'G'}): -2.0,\n",
    "            frozenset({'L', 'Y'}): -1.0, frozenset({'P', 'K'}): -1.0, frozenset({'R', 'M'}): -1.0,\n",
    "            frozenset({'R', 'K'}): 2.0, frozenset({'W', 'M'}): -1.0, frozenset({'C'}): 9.0, frozenset({'H'}): 8.0,\n",
    "            frozenset({'I', 'S'}): -2.0, frozenset({'S', 'B'}): 0.0, frozenset({'X', 'G'}): -1.0,\n",
    "            frozenset({'V', 'X'}): -1.0, frozenset({'L', 'G'}): -4.0, frozenset({'T', 'C'}): -1.0,\n",
    "            frozenset({'T'}): 5.0, frozenset({'Q', 'E'}): 2.0, frozenset({'B'}): 4.0, frozenset({'P'}): 7.0,\n",
    "            frozenset({'W', 'K'}): -3.0, frozenset({'I'}): 4.0, frozenset({'*'}): 1.0, frozenset({'R', 'B'}): -1.0,\n",
    "            frozenset({'T', 'F'}): -2.0, frozenset({'T', 'M'}): -1.0, frozenset({'Z', 'N'}): 0.0,\n",
    "            frozenset({'I', 'E'}): -3.0, frozenset({'T', 'N'}): 0.0, frozenset({'S', 'F'}): -2.0,\n",
    "            frozenset({'E', 'B'}): 1.0, frozenset({'Y', 'A'}): -2.0, frozenset({'W', 'G'}): -2.0,\n",
    "            frozenset({'I', 'R'}): -3.0, frozenset({'R', 'E'}): 0.0, frozenset({'Y'}): 7.0, frozenset({'F', 'N'}): -3.0,\n",
    "            frozenset({'T', 'X'}): 0.0, frozenset({'G', 'M'}): -3.0, frozenset({'K', 'G'}): -2.0,\n",
    "            frozenset({'S', 'C'}): -1.0, frozenset({'Z', 'C'}): -3.0, frozenset({'H', 'Z'}): 0.0,\n",
    "            frozenset({'Q', 'B'}): 0.0, frozenset({'C', 'V'}): -1.0, frozenset({'V'}): 4.0, frozenset({'I', 'A'}): -1.0,\n",
    "            frozenset({'H', 'Y'}): 2.0, frozenset({'W', 'X'}): -2.0, frozenset({'Q', 'N'}): 0.0,\n",
    "            frozenset({'R', 'N'}): 0.0, frozenset({'B', 'D'}): 4.0, frozenset({'H', 'X'}): -1.0,\n",
    "            frozenset({'C', 'X'}): -2.0, frozenset({'*', 'A'}): -4.0, frozenset({'N', 'B'}): 3.0,\n",
    "            frozenset({'B', 'G'}): -1.0, frozenset({'*', 'P'}): -4.0, frozenset({'K', 'Y'}): -2.0,\n",
    "            frozenset({'Y', 'M'}): -1.0, frozenset({'P', 'L'}): -3.0, frozenset({'Q', 'K'}): 1.0,\n",
    "            frozenset({'S', 'V'}): -2.0, frozenset({'F'}): 6.0, frozenset({'M', 'A'}): -1.0,\n",
    "            frozenset({'F', 'A'}): -2.0, frozenset({'N', 'X'}): -1.0, frozenset({'T', 'P'}): -1.0,\n",
    "            frozenset({'S', 'N'}): 1.0, frozenset({'Q', 'G'}): -2.0, frozenset({'B', 'A'}): -2.0,\n",
    "            frozenset({'H', 'P'}): -2.0, frozenset({'C', 'P'}): -3.0, frozenset({'X', 'K'}): -1.0,\n",
    "            frozenset({'Q', 'M'}): 0.0, frozenset({'Z', 'P'}): -1.0, frozenset({'B', 'M'}): -3.0,\n",
    "            frozenset({'I', 'L'}): 2.0, frozenset({'H', 'T'}): -2.0}\n",
    "\n",
    "\n",
    "class MultipleAlignment:\n",
    "    \"\"\"\n",
    "    represents a simple multiple alignment\n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(self):\n",
    "        self.sequence_matrix = []\n",
    "\n",
    "    def add_sequence(self, sequence):\n",
    "        \"\"\"\n",
    "        adds sequence characters to matrix, appending\n",
    "        to existing values such that aligned characters\n",
    "        will end up occupying the same index\n",
    "        ex:\n",
    "        'ZWAB'\n",
    "        'B-ZX'\n",
    "        will end up as\n",
    "        ['ZB','W-','AZ','BX']\n",
    "        :param sequence:\n",
    "        \"\"\"\n",
    "        for idx, char in enumerate(sequence):\n",
    "            if idx < len(self.sequence_matrix):\n",
    "                self.sequence_matrix[idx] += char\n",
    "            else:\n",
    "                self.sequence_matrix.append(char)\n",
    "\n",
    "    def sum_aligned_char_scores(self, scoring_function, *sf_args):\n",
    "        \"\"\"\n",
    "        addition commutes. aligned characters of matrix\n",
    "        are summed, simulating the summation of whichever\n",
    "        scoring function\n",
    "        :param scoring_function:\n",
    "        :param sf_args:\n",
    "        :return:\n",
    "        \"\"\"\n",
    "        score_sum = 0.0\n",
    "        for aligned_chars in self.sequence_matrix:\n",
    "            if all(char not in ILLEGAL_CHARS for char in aligned_chars):\n",
    "                score_sum += scoring_function(aligned_chars, sf_args)\n",
    "        return score_sum\n",
    "\n",
    "    def minimize_aligned_char_scores(self, scoring_function, *sf_args):\n",
    "        \"\"\"\n",
    "        returns the minimum value returned by the\n",
    "        scoring function for aligned characters\n",
    "        :param scoring_function:\n",
    "        :param sf_args:\n",
    "        :return:\n",
    "        \"\"\"\n",
    "        score_list = []\n",
    "        for aligned_chars in self.sequence_matrix:\n",
    "            score_list.append(scoring_function(aligned_chars, sf_args))\n",
    "        return min(score_list)\n",
    "\n",
    "def sum_of_pairs(aligned_chars, param_tuple):\n",
    "    \"\"\"Sum-of-Pairs (SP-score)\n",
    "    Compeau, P, Pevzner, P,c Active Learning Publishers,\n",
    "    Bioinformatics Algorithms: An Active Learning Approach,\n",
    "    2nd Ed., Vol. I, p. 290\n",
    "    :param aligned_chars: the characters in the alignment\n",
    "    :param param_tuple: a tuple with a scoring matrix wrapper\n",
    "    \"\"\"\n",
    "    matrix_wrapper = param_tuple[0]\n",
    "    sp = 0.0\n",
    "    for i in range(0, len(aligned_chars)):\n",
    "        for j in range(i + 1, len(aligned_chars)):\n",
    "            sp += matrix_wrapper(aligned_chars[i], aligned_chars[j])\n",
    "    return sp + 0  # trick from http://goo.gl/8u8kQw\n",
    "\n",
    "def h(char_i, char_j):\n",
    "    \"\"\"\n",
    "    simply wraps dictionary\n",
    "    \"\"\"\n",
    "    return BLOSUM62[frozenset([char_i, char_j])]\n",
    "\n",
    "def set_globs(s1,s2):\n",
    "    global _s1\n",
    "    global _s2\n",
    "    global _dp\n",
    "    _s1 = '*' + s1\n",
    "    _s2 = '*' + s2\n",
    "\n",
    "    # create DP matrix with S1 + 1 columns and S2 + 1 rows\n",
    "    _dp = [[0 for x in range(len(_s1) + 1)] for x in range(len(_s2) + 1)]\n",
    "\n",
    "def fill_l_dp_matrix():\n",
    "    global _dp\n",
    "    for i in range(len(_s2)):\n",
    "        for j in range(len(_s1)):\n",
    "            if i > 0 and j > 0:\n",
    "                _dp[i][j] = max(_dp[i - 1][j - 1] + h(_s2[i], _s1[j]),\n",
    "                                _dp[i - 1][j] + GAP_PENALTY,\n",
    "                                _dp[i][j - 1] + GAP_PENALTY,\n",
    "                                0) # allow for local alignment\n",
    "            elif i > 0:\n",
    "                _dp[i][j] = _dp[i - 1][j] + GAP_PENALTY\n",
    "            elif j > 0:\n",
    "                _dp[i][j] = _dp[i][j - 1] + GAP_PENALTY\n",
    "            else:\n",
    "                _dp[i][j] = 0\n",
    "\n",
    "def fill_dp_matrix():\n",
    "    global _dp\n",
    "    for i in range(len(_s2)):\n",
    "        for j in range(len(_s1)):\n",
    "            if i > 0 and j > 0:\n",
    "                _dp[i][j] = max(_dp[i - 1][j - 1] + h(_s2[i], _s1[j]),\n",
    "                                _dp[i - 1][j] + GAP_PENALTY,\n",
    "                                _dp[i][j - 1] + GAP_PENALTY)\n",
    "            elif i > 0:\n",
    "                _dp[i][j] = _dp[i - 1][j] + GAP_PENALTY\n",
    "            elif j > 0:\n",
    "                _dp[i][j] = _dp[i][j - 1] + GAP_PENALTY\n",
    "            else:\n",
    "                _dp[i][j] = 0\n",
    "\n",
    "def print_dp_matrix():\n",
    "    for c in range(len(_s1)):\n",
    "        print(' {0}'.format(_s1[c]).center(4, ' '), end='')\n",
    "    print()\n",
    "    for i in range(len(_s2)):\n",
    "        print('{0}'.format(_s2[i]), end='')\n",
    "        for j in range(len(_s1)):\n",
    "            print('{0}'.format(_dp[i][j]).center(4, ' '), end='')\n",
    "        print()\n",
    "\n",
    "def lbacktrace():\n",
    "    s1_alignment = ''\n",
    "    s2_alignment = ''\n",
    "    i = len(_s2) - 1\n",
    "    j = len(_s1) - 1\n",
    "    while i > 0 or j > 0:\n",
    "        if i > 0 and j > 0:\n",
    "            diag = _dp[i - 1][j - 1]\n",
    "            up = _dp[i - 1][j]\n",
    "            left = _dp[i][j - 1]\n",
    "            if _dp[i][j] == 0:\n",
    "                if len(s2_alignment.strip()) > 0: #done with local alignment, abort\n",
    "                    i = 0\n",
    "                else:\n",
    "                    s1_alignment = _s1[j] + s1_alignment\n",
    "                    s2_alignment = ' ' + s2_alignment\n",
    "                    i -= 1\n",
    "                    j -= 1\n",
    "            elif (_dp[i][j] - diag) == h(_s2[i],_s1[j]):\n",
    "                s1_alignment = _s1[j] + s1_alignment\n",
    "                s2_alignment = _s2[i] + s2_alignment\n",
    "                i -= 1\n",
    "                j -= 1\n",
    "            elif (_dp[i][j] - up) == GAP_PENALTY:\n",
    "                s1_alignment = '*' + s1_alignment\n",
    "                s2_alignment = _s2[i] + s2_alignment\n",
    "                i -= 1\n",
    "            elif (_dp[i][j] - left) == GAP_PENALTY:\n",
    "                s1_alignment = _s1[j] + s1_alignment\n",
    "                s2_alignment = '*' + s2_alignment\n",
    "                j -= 1\n",
    "        elif i > 0:  # up\n",
    "            s1_alignment = '*' + s1_alignment\n",
    "            s2_alignment = _s2[i] + s2_alignment\n",
    "            i -= 1\n",
    "        elif j > 0:  # left\n",
    "            s1_alignment = _s1[j] + s1_alignment\n",
    "            s2_alignment = ' ' + s2_alignment\n",
    "            j -= 1\n",
    "    return s1_alignment, s2_alignment\n",
    "\n",
    "def backtrace():\n",
    "    s1_alignment = ''\n",
    "    s2_alignment = ''\n",
    "    i = len(_s2) - 1\n",
    "    j = len(_s1) - 1\n",
    "    while i > 0 or j > 0:\n",
    "        if i > 0 and j > 0:\n",
    "            diag = _dp[i - 1][j - 1]\n",
    "            up = _dp[i - 1][j]\n",
    "            left = _dp[i][j - 1]\n",
    "            if (_dp[i][j] - up) == GAP_PENALTY:\n",
    "                s1_alignment = '*' + s1_alignment\n",
    "                s2_alignment = _s2[i] + s2_alignment\n",
    "                i -= 1\n",
    "            elif (_dp[i][j] - diag) == h(_s2[i],_s1[j]):\n",
    "                s1_alignment = _s1[j] + s1_alignment\n",
    "                s2_alignment = _s2[i] + s2_alignment\n",
    "                i -= 1\n",
    "                j -= 1\n",
    "            elif (_dp[i][j] - left) == GAP_PENALTY:\n",
    "                s1_alignment = _s1[j] + s1_alignment\n",
    "                s2_alignment = '*' + s2_alignment\n",
    "                j -= 1\n",
    "        elif i > 0:  # up\n",
    "            s1_alignment = '*' + s1_alignment\n",
    "            s2_alignment = _s2[i] + s2_alignment\n",
    "            i -= 1\n",
    "        elif j > 0:  # left\n",
    "            s1_alignment = _s1[j] + s1_alignment\n",
    "            s2_alignment = '*' + s2_alignment\n",
    "            j -= 1\n",
    "    return s1_alignment, s2_alignment\n",
    "\n",
    "def lalign(s1,s2):\n",
    "    set_globs(s1.upper(),s2.upper())\n",
    "    print('DP matrix:')\n",
    "    fill_l_dp_matrix()\n",
    "    print_dp_matrix()\n",
    "    return lbacktrace()\n",
    "\n",
    "def galign(s1,s2):\n",
    "    set_globs(s1.upper(),s2.upper())\n",
    "    print('DP matrix:')\n",
    "    fill_dp_matrix()\n",
    "    print_dp_matrix()\n",
    "    return backtrace()\n",
    "\n",
    "r = 'BMPTSQCVNS'\n",
    "l = 'vvtTSQalxc'\n",
    "\n",
    "lalign1,lalign2 = lalign(r,l)\n",
    "print(lalign1+'\\n'+lalign2)\n",
    "ma = MultipleAlignment()\n",
    "ma.add_sequence(lalign1.upper())\n",
    "ma.add_sequence(lalign2.upper())\n",
    "print('Sum of Pairs = {0}'.format(\n",
    "        ma.sum_aligned_char_scores(sum_of_pairs, h)))\n",
    "\n",
    "galign1,galign2 = galign(r,l)\n",
    "print(galign1+'\\n'+galign2)\n",
    "gma = MultipleAlignment()\n",
    "gma.add_sequence(galign1.upper())\n",
    "gma.add_sequence(galign2.upper())\n",
    "print('Sum of Pairs = {0}'.format(\n",
    "        gma.sum_aligned_char_scores(sum_of_pairs, h)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### a. 4) Discuss the alignment"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### b. 1) a better global alignment as compared to a local alignment (using blosum62 scoring matrix), both 10 amino acids long."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "r: INHBRADLCD\n",
      "g: IhHpRwDiCp\n"
     ]
    }
   ],
   "source": [
    "from random import randint\n",
    "\n",
    "SIGMA = 'ARNDCQEGHILKMFPSTWYVBZX'\n",
    "\n",
    "def r_char():\n",
    "    \"\"\"\n",
    "    Generates a random character\n",
    "    :return: a random character from SIGMA\n",
    "    \"\"\"\n",
    "    return SIGMA[randint(0,len(SIGMA)-1)]\n",
    "\n",
    "\n",
    "def random_sequence(l):\n",
    "    \"\"\"\n",
    "    Generates a random sequence\n",
    "    :param l: length of random sequence to build\n",
    "    :return: random sequence of length l\n",
    "    \"\"\"\n",
    "    r_seq = ''\n",
    "    for i in range(l):\n",
    "        r_seq += r_char()\n",
    "    return r_seq\n",
    "\n",
    "def global_align_sequence(r_seq):\n",
    "    \"\"\"\n",
    "    Matches every other character in r_seq\n",
    "    :param r_seq: any random sequence\n",
    "    :return: a sequence that will align globally to r_seq\n",
    "    \"\"\"\n",
    "    g_seq = ''\n",
    "    for i in range(len(r_seq)):\n",
    "        if i%2 == 0:\n",
    "            g_seq += r_seq[i]\n",
    "        else:\n",
    "            g_seq += r_char().lower()\n",
    "    return g_seq\n",
    "\n",
    "r = random_sequence(10)\n",
    "g = global_align_sequence(r)\n",
    "\n",
    "print('r: ' + r)\n",
    "print('g: ' + g)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### b. 2) Discuss criteria and any limitations in generating your sequences."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### b. 3) Align the two sequences using dynamic programming using the blosum62 scoring matrix. Show the DP matrix and the resultant alignment."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DP matrix:\n",
      "  *   I   N   H   B   R   A   D   L   C   D \n",
      "* 0   -4  -8 -12 -16 -20 -24 -28 -32 -36 -40 \n",
      "I -4 4.0 0.0  0   0   0   0   0   0   0   0  \n",
      "H -8 0.0 5.0 8.0 4.0 0.0  0   0   0   0   0  \n",
      "H-12  0  1.0 13.09.0 5.0 1.0  0   0   0   0  \n",
      "P-16  0   0  9.0 11.07.0 4.0 0.0  0   0   0  \n",
      "R-20  0  0.0 5.0 8.0 16.012.08.0 4.0 0.0  0  \n",
      "W-24  0   0  1.0 4.0 12.013.09.0 6.0 2.0  0  \n",
      "D-28  0  1.0  0  5.0 8.0 10.019.015.011.08.0 \n",
      "I-32  0   0   0  1.0 4.0 7.0 15.021.017.013.0\n",
      "C-36  0   0   0   0  0.0 4.0 11.017.030.026.0\n",
      "P-40  0   0   0   0   0  0.0 7.0 13.026.029.0\n",
      "INHBRADLCD\n",
      "IHHPRWDICP\n",
      "Sum of Pairs = 29.0\n",
      "DP matrix:\n",
      "  *   I   N   H   B   R   A   D   L   C   D \n",
      "* 0   -4  -8 -12 -16 -20 -24 -28 -32 -36 -40 \n",
      "I -4 4.0 0.0 -4.0-8.0-12.0-16.0-20.0-24.0-28.0-32.0\n",
      "H -8 0.0 5.0 8.0 4.0 0.0 -4.0-8.0-12.0-16.0-20.0\n",
      "H-12 -4.01.0 13.09.0 5.0 1.0 -3.0-7.0-11.0-15.0\n",
      "P-16 -8.0-3.09.0 11.07.0 4.0 0.0 -4.0-8.0-12.0\n",
      "R-20 -12.0-7.05.0 8.0 16.012.08.0 4.0 0.0 -4.0\n",
      "W-24 -16.0-11.01.0 4.0 12.013.09.0 6.0 2.0 -2.0\n",
      "D-28 -20.0-15.0-3.05.0 8.0 10.019.015.011.08.0 \n",
      "I-32 -24.0-19.0-7.01.0 4.0 7.0 15.021.017.013.0\n",
      "C-36 -28.0-23.0-11.0-3.00.0 4.0 11.017.030.026.0\n",
      "P-40 -32.0-27.0-15.0-7.0-4.00.0 7.0 13.026.029.0\n",
      "INHBRADLCD\n",
      "IHHPRWDICP\n",
      "Sum of Pairs = 29.0\n"
     ]
    }
   ],
   "source": [
    "import math\n",
    "from random import randint\n",
    "\n",
    "SIGMA = 'ARNDCQEGHILKMFPSTWYVBZX'\n",
    "ILLEGAL_CHARS = ' -\\n'\n",
    "GAP_PENALTY = -4 # taken from blosum62 matrix\n",
    "BLOSUM62 = {frozenset({'R', 'P'}): -2.0, frozenset({'S', 'G'}): 0.0, frozenset({'T', 'K'}): -1.0,\n",
    "            frozenset({'N', 'G'}): 0.0, frozenset({'Z', 'G'}): -2.0, frozenset({'P', 'B'}): -2.0,\n",
    "            frozenset({'T', 'W'}): -2.0, frozenset({'N', 'D'}): 1.0, frozenset({'*', 'N'}): -4.0, frozenset({'L'}): 4.0,\n",
    "            frozenset({'Q', 'A'}): -1.0, frozenset({'N', 'K'}): 0.0, frozenset({'*', 'M'}): -4.0,\n",
    "            frozenset({'N', 'A'}): -2.0, frozenset({'Q', 'P'}): -1.0, frozenset({'X', 'A'}): 0.0,\n",
    "            frozenset({'Z', 'K'}): 1.0, frozenset({'*', 'K'}): -4.0, frozenset({'H', 'D'}): -1.0,\n",
    "            frozenset({'C', 'D'}): -3.0, frozenset({'B', 'X'}): -1.0, frozenset({'Q', 'R'}): 1.0,\n",
    "            frozenset({'W', 'P'}): -4.0, frozenset({'S', 'Y'}): -2.0, frozenset({'*', 'W'}): -4.0,\n",
    "            frozenset({'W', 'D'}): -4.0, frozenset({'K', 'M'}): -1.0, frozenset({'I', 'V'}): 3.0,\n",
    "            frozenset({'E', 'K'}): 1.0, frozenset({'C', 'Y'}): -2.0, frozenset({'R', 'D'}): -2.0,\n",
    "            frozenset({'Z', 'B'}): 1.0, frozenset({'P', 'G'}): -2.0, frozenset({'Q', 'C'}): -3.0,\n",
    "            frozenset({'R', 'S'}): -1.0, frozenset({'B', 'L'}): -4.0, frozenset({'R', 'Z'}): 0.0,\n",
    "            frozenset({'I', 'F'}): 0.0, frozenset({'S', 'A'}): 1.0, frozenset({'I', 'B'}): -3.0,\n",
    "            frozenset({'F', 'Y'}): 3.0, frozenset({'D', 'A'}): -2.0, frozenset({'E', 'C'}): -4.0,\n",
    "            frozenset({'H', 'E'}): 0.0, frozenset({'I', 'C'}): -1.0, frozenset({'H', 'I'}): -3.0,\n",
    "            frozenset({'Q', 'T'}): -1.0, frozenset({'V', 'D'}): -3.0, frozenset({'K', 'D'}): -1.0,\n",
    "            frozenset({'Q', 'V'}): -2.0, frozenset({'C', 'B'}): -3.0, frozenset({'*', 'V'}): -4.0,\n",
    "            frozenset({'X', 'Y'}): -1.0, frozenset({'F', 'V'}): -1.0, frozenset({'I', 'M'}): 1.0,\n",
    "            frozenset({'P', 'M'}): -2.0, frozenset({'E', 'W'}): -3.0, frozenset({'W', 'A'}): -3.0,\n",
    "            frozenset({'E', 'L'}): -3.0, frozenset({'T', 'Z'}): -1.0, frozenset({'*', 'E'}): -4.0,\n",
    "            frozenset({'I', 'N'}): -3.0, frozenset({'E', 'Z'}): 4.0, frozenset({'H', 'A'}): -2.0,\n",
    "            frozenset({'B', 'K'}): 0.0, frozenset({'M'}): 5.0, frozenset({'T', 'Y'}): -2.0, frozenset({'D', 'G'}): -1.0,\n",
    "            frozenset({'K'}): 5.0, frozenset({'W', 'N'}): -4.0, frozenset({'I', 'X'}): -1.0, frozenset({'A'}): 4.0,\n",
    "            frozenset({'N', 'L'}): -3.0, frozenset({'W', 'C'}): -2.0, frozenset({'H', 'W'}): -2.0,\n",
    "            frozenset({'F', 'M'}): 0.0, frozenset({'S', 'D'}): 0.0, frozenset({'*', 'D'}): -4.0,\n",
    "            frozenset({'P', 'X'}): -2.0, frozenset({'T', 'L'}): -1.0, frozenset({'L', 'M'}): 2.0,\n",
    "            frozenset({'D', 'M'}): -3.0, frozenset({'*', 'T'}): -4.0, frozenset({'P', 'Y'}): -3.0,\n",
    "            frozenset({'T', 'G'}): -2.0, frozenset({'R', 'Y'}): -2.0, frozenset({'I', 'P'}): -3.0,\n",
    "            frozenset({'*', 'S'}): -4.0, frozenset({'S', 'W'}): -3.0, frozenset({'F', 'G'}): -3.0,\n",
    "            frozenset({'I', 'T'}): -1.0, frozenset({'T', 'E'}): -1.0, frozenset({'G'}): 6.0, frozenset({'H', 'N'}): 1.0,\n",
    "            frozenset({'W', 'Y'}): 2.0, frozenset({'V', 'M'}): 1.0, frozenset({'N', 'C'}): -3.0,\n",
    "            frozenset({'V', 'L'}): 1.0, frozenset({'Q', 'Z'}): 3.0, frozenset({'R', 'G'}): -2.0,\n",
    "            frozenset({'F', 'X'}): -1.0, frozenset({'W', 'B'}): -4.0, frozenset({'H', 'B'}): 0.0,\n",
    "            frozenset({'V', 'A'}): 0.0, frozenset({'Z', 'Y'}): -2.0, frozenset({'W', 'L'}): -2.0,\n",
    "            frozenset({'K', 'A'}): -1.0, frozenset({'T', 'A'}): 0.0, frozenset({'Y', 'D'}): -3.0,\n",
    "            frozenset({'R', 'C'}): -3.0, frozenset({'X', 'M'}): -1.0, frozenset({'Q', 'L'}): -2.0,\n",
    "            frozenset({'Q', 'D'}): 0.0, frozenset({'I', 'W'}): -3.0, frozenset({'E', 'D'}): 2.0,\n",
    "            frozenset({'R', 'T'}): -1.0, frozenset({'Y', 'G'}): -3.0, frozenset({'S', 'E'}): 0.0,\n",
    "            frozenset({'F', 'P'}): -4.0, frozenset({'S'}): 4.0, frozenset({'W'}): 11.0, frozenset({'C', 'A'}): 0.0,\n",
    "            frozenset({'T', 'B'}): -1.0, frozenset({'R', 'F'}): -3.0, frozenset({'H', 'K'}): -1.0,\n",
    "            frozenset({'H', 'M'}): -2.0, frozenset({'C', 'M'}): -1.0, frozenset({'S', 'M'}): -1.0,\n",
    "            frozenset({'Q', 'I'}): -3.0, frozenset({'R'}): 5.0, frozenset({'F', 'C'}): -2.0,\n",
    "            frozenset({'H', 'F'}): -1.0, frozenset({'V', 'G'}): -3.0, frozenset({'Z', 'L'}): -3.0,\n",
    "            frozenset({'W', 'V'}): -3.0, frozenset({'*', 'Z'}): -4.0, frozenset({'H', 'C'}): -3.0,\n",
    "            frozenset({'N', 'V'}): -3.0, frozenset({'L', 'D'}): -4.0, frozenset({'R', 'V'}): -3.0,\n",
    "            frozenset({'Q', 'Y'}): -1.0, frozenset({'S', 'L'}): -2.0, frozenset({'N', 'M'}): -2.0,\n",
    "            frozenset({'Z', 'M'}): -1.0, frozenset({'F', 'L'}): 0.0, frozenset({'H', 'G'}): -2.0,\n",
    "            frozenset({'C', 'G'}): -3.0, frozenset({'F', 'Z'}): -3.0, frozenset({'V', 'B'}): -3.0,\n",
    "            frozenset({'H', 'L'}): -3.0, frozenset({'E'}): 5.0, frozenset({'*', 'C'}): -4.0,\n",
    "            frozenset({'H', '*'}): -4.0, frozenset({'F', 'B'}): -3.0, frozenset({'E', 'A'}): -1.0,\n",
    "            frozenset({'R', 'A'}): -1.0, frozenset({'H', 'R'}): 0.0, frozenset({'E', 'N'}): 0.0,\n",
    "            frozenset({'T', 'V'}): 0.0, frozenset({'*', 'B'}): -4.0, frozenset({'G', 'A'}): 0.0,\n",
    "            frozenset({'H', 'S'}): -1.0, frozenset({'Q', 'F'}): -3.0, frozenset({'N', 'P'}): -2.0,\n",
    "            frozenset({'D'}): 6.0, frozenset({'Z', 'X'}): -1.0, frozenset({'F', 'W'}): 1.0, frozenset({'N', 'Y'}): -2.0,\n",
    "            frozenset({'X', 'D'}): -1.0, frozenset({'Z', 'V'}): -2.0, frozenset({'Q', '*'}): -4.0,\n",
    "            frozenset({'Q', 'S'}): 0.0, frozenset({'L', 'K'}): -2.0, frozenset({'W', 'Z'}): -3.0,\n",
    "            frozenset({'*', 'Y'}): -4.0, frozenset({'*', 'L'}): -4.0, frozenset({'Q'}): 5.0,\n",
    "            frozenset({'L', 'A'}): -1.0, frozenset({'E', 'V'}): -2.0, frozenset({'I', 'Z'}): -3.0,\n",
    "            frozenset({'I', 'K'}): -3.0, frozenset({'E', 'Y'}): -2.0, frozenset({'C', 'K'}): -3.0,\n",
    "            frozenset({'I', 'Y'}): -1.0, frozenset({'*', 'G'}): -4.0, frozenset({'E', 'X'}): -1.0,\n",
    "            frozenset({'S', 'X'}): 0.0, frozenset({'Q', 'X'}): -1.0, frozenset({'P', 'A'}): -1.0,\n",
    "            frozenset({'F', 'K'}): -3.0, frozenset({'S', 'Z'}): 0.0, frozenset({'Z'}): 4.0, frozenset({'I', 'D'}): -3.0,\n",
    "            frozenset({'Q', 'H'}): 0.0, frozenset({'*', 'F'}): -4.0, frozenset({'F', 'D'}): -3.0,\n",
    "            frozenset({'T', 'D'}): -1.0, frozenset({'S', 'K'}): 0.0, frozenset({'I', '*'}): -4.0,\n",
    "            frozenset({'S', 'T'}): 1.0, frozenset({'C', 'L'}): -1.0, frozenset({'E', 'P'}): -1.0,\n",
    "            frozenset({'X'}): -1.0, frozenset({'R', 'X'}): -1.0, frozenset({'B', 'Y'}): -3.0,\n",
    "            frozenset({'R', 'L'}): -2.0, frozenset({'E', 'M'}): -2.0, frozenset({'*', 'R'}): -4.0,\n",
    "            frozenset({'H', 'V'}): -3.0, frozenset({'Q', 'W'}): -2.0, frozenset({'E', 'F'}): -3.0,\n",
    "            frozenset({'Z', 'D'}): 1.0, frozenset({'P', 'V'}): -2.0, frozenset({'V', 'K'}): -2.0, frozenset({'N'}): 6.0,\n",
    "            frozenset({'R', 'W'}): -3.0, frozenset({'S', 'P'}): -1.0, frozenset({'Z', 'A'}): -1.0,\n",
    "            frozenset({'P', 'D'}): -1.0, frozenset({'*', 'X'}): -4.0, frozenset({'V', 'Y'}): -1.0,\n",
    "            frozenset({'X', 'L'}): -1.0, frozenset({'I', 'G'}): -4.0, frozenset({'E', 'G'}): -2.0,\n",
    "            frozenset({'L', 'Y'}): -1.0, frozenset({'P', 'K'}): -1.0, frozenset({'R', 'M'}): -1.0,\n",
    "            frozenset({'R', 'K'}): 2.0, frozenset({'W', 'M'}): -1.0, frozenset({'C'}): 9.0, frozenset({'H'}): 8.0,\n",
    "            frozenset({'I', 'S'}): -2.0, frozenset({'S', 'B'}): 0.0, frozenset({'X', 'G'}): -1.0,\n",
    "            frozenset({'V', 'X'}): -1.0, frozenset({'L', 'G'}): -4.0, frozenset({'T', 'C'}): -1.0,\n",
    "            frozenset({'T'}): 5.0, frozenset({'Q', 'E'}): 2.0, frozenset({'B'}): 4.0, frozenset({'P'}): 7.0,\n",
    "            frozenset({'W', 'K'}): -3.0, frozenset({'I'}): 4.0, frozenset({'*'}): 1.0, frozenset({'R', 'B'}): -1.0,\n",
    "            frozenset({'T', 'F'}): -2.0, frozenset({'T', 'M'}): -1.0, frozenset({'Z', 'N'}): 0.0,\n",
    "            frozenset({'I', 'E'}): -3.0, frozenset({'T', 'N'}): 0.0, frozenset({'S', 'F'}): -2.0,\n",
    "            frozenset({'E', 'B'}): 1.0, frozenset({'Y', 'A'}): -2.0, frozenset({'W', 'G'}): -2.0,\n",
    "            frozenset({'I', 'R'}): -3.0, frozenset({'R', 'E'}): 0.0, frozenset({'Y'}): 7.0, frozenset({'F', 'N'}): -3.0,\n",
    "            frozenset({'T', 'X'}): 0.0, frozenset({'G', 'M'}): -3.0, frozenset({'K', 'G'}): -2.0,\n",
    "            frozenset({'S', 'C'}): -1.0, frozenset({'Z', 'C'}): -3.0, frozenset({'H', 'Z'}): 0.0,\n",
    "            frozenset({'Q', 'B'}): 0.0, frozenset({'C', 'V'}): -1.0, frozenset({'V'}): 4.0, frozenset({'I', 'A'}): -1.0,\n",
    "            frozenset({'H', 'Y'}): 2.0, frozenset({'W', 'X'}): -2.0, frozenset({'Q', 'N'}): 0.0,\n",
    "            frozenset({'R', 'N'}): 0.0, frozenset({'B', 'D'}): 4.0, frozenset({'H', 'X'}): -1.0,\n",
    "            frozenset({'C', 'X'}): -2.0, frozenset({'*', 'A'}): -4.0, frozenset({'N', 'B'}): 3.0,\n",
    "            frozenset({'B', 'G'}): -1.0, frozenset({'*', 'P'}): -4.0, frozenset({'K', 'Y'}): -2.0,\n",
    "            frozenset({'Y', 'M'}): -1.0, frozenset({'P', 'L'}): -3.0, frozenset({'Q', 'K'}): 1.0,\n",
    "            frozenset({'S', 'V'}): -2.0, frozenset({'F'}): 6.0, frozenset({'M', 'A'}): -1.0,\n",
    "            frozenset({'F', 'A'}): -2.0, frozenset({'N', 'X'}): -1.0, frozenset({'T', 'P'}): -1.0,\n",
    "            frozenset({'S', 'N'}): 1.0, frozenset({'Q', 'G'}): -2.0, frozenset({'B', 'A'}): -2.0,\n",
    "            frozenset({'H', 'P'}): -2.0, frozenset({'C', 'P'}): -3.0, frozenset({'X', 'K'}): -1.0,\n",
    "            frozenset({'Q', 'M'}): 0.0, frozenset({'Z', 'P'}): -1.0, frozenset({'B', 'M'}): -3.0,\n",
    "            frozenset({'I', 'L'}): 2.0, frozenset({'H', 'T'}): -2.0}\n",
    "\n",
    "\n",
    "class MultipleAlignment:\n",
    "    \"\"\"\n",
    "    represents a simple multiple alignment\n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(self):\n",
    "        self.sequence_matrix = []\n",
    "\n",
    "    def add_sequence(self, sequence):\n",
    "        \"\"\"\n",
    "        adds sequence characters to matrix, appending\n",
    "        to existing values such that aligned characters\n",
    "        will end up occupying the same index\n",
    "        ex:\n",
    "        'ZWAB'\n",
    "        'B-ZX'\n",
    "        will end up as\n",
    "        ['ZB','W-','AZ','BX']\n",
    "        :param sequence:\n",
    "        \"\"\"\n",
    "        for idx, char in enumerate(sequence):\n",
    "            if idx < len(self.sequence_matrix):\n",
    "                self.sequence_matrix[idx] += char\n",
    "            else:\n",
    "                self.sequence_matrix.append(char)\n",
    "\n",
    "    def sum_aligned_char_scores(self, scoring_function, *sf_args):\n",
    "        \"\"\"\n",
    "        addition commutes. aligned characters of matrix\n",
    "        are summed, simulating the summation of whichever\n",
    "        scoring function\n",
    "        :param scoring_function:\n",
    "        :param sf_args:\n",
    "        :return:\n",
    "        \"\"\"\n",
    "        score_sum = 0.0\n",
    "        for aligned_chars in self.sequence_matrix:\n",
    "            if all(char not in ILLEGAL_CHARS for char in aligned_chars):\n",
    "                score_sum += scoring_function(aligned_chars, sf_args)\n",
    "        return score_sum\n",
    "\n",
    "    def minimize_aligned_char_scores(self, scoring_function, *sf_args):\n",
    "        \"\"\"\n",
    "        returns the minimum value returned by the\n",
    "        scoring function for aligned characters\n",
    "        :param scoring_function:\n",
    "        :param sf_args:\n",
    "        :return:\n",
    "        \"\"\"\n",
    "        score_list = []\n",
    "        for aligned_chars in self.sequence_matrix:\n",
    "            score_list.append(scoring_function(aligned_chars, sf_args))\n",
    "        return min(score_list)\n",
    "\n",
    "def sum_of_pairs(aligned_chars, param_tuple):\n",
    "    \"\"\"Sum-of-Pairs (SP-score)\n",
    "    Compeau, P, Pevzner, P,c Active Learning Publishers,\n",
    "    Bioinformatics Algorithms: An Active Learning Approach,\n",
    "    2nd Ed., Vol. I, p. 290\n",
    "    :param aligned_chars: the characters in the alignment\n",
    "    :param param_tuple: a tuple with a scoring matrix wrapper\n",
    "    \"\"\"\n",
    "    matrix_wrapper = param_tuple[0]\n",
    "    sp = 0.0\n",
    "    for i in range(0, len(aligned_chars)):\n",
    "        for j in range(i + 1, len(aligned_chars)):\n",
    "            sp += matrix_wrapper(aligned_chars[i], aligned_chars[j])\n",
    "    return sp + 0  # trick from http://goo.gl/8u8kQw\n",
    "\n",
    "def h(char_i, char_j):\n",
    "    \"\"\"\n",
    "    simply wraps dictionary\n",
    "    \"\"\"\n",
    "    return BLOSUM62[frozenset([char_i, char_j])]\n",
    "\n",
    "def set_globs(s1,s2):\n",
    "    global _s1\n",
    "    global _s2\n",
    "    global _dp\n",
    "    _s1 = '*' + s1\n",
    "    _s2 = '*' + s2\n",
    "\n",
    "    # create DP matrix with S1 + 1 columns and S2 + 1 rows\n",
    "    _dp = [[0 for x in range(len(_s1) + 1)] for x in range(len(_s2) + 1)]\n",
    "\n",
    "def fill_l_dp_matrix():\n",
    "    global _dp\n",
    "    for i in range(len(_s2)):\n",
    "        for j in range(len(_s1)):\n",
    "            if i > 0 and j > 0:\n",
    "                _dp[i][j] = max(_dp[i - 1][j - 1] + h(_s2[i], _s1[j]),\n",
    "                                _dp[i - 1][j] + GAP_PENALTY,\n",
    "                                _dp[i][j - 1] + GAP_PENALTY,\n",
    "                                0) # allow for local alignment\n",
    "            elif i > 0:\n",
    "                _dp[i][j] = _dp[i - 1][j] + GAP_PENALTY\n",
    "            elif j > 0:\n",
    "                _dp[i][j] = _dp[i][j - 1] + GAP_PENALTY\n",
    "            else:\n",
    "                _dp[i][j] = 0\n",
    "\n",
    "def fill_dp_matrix():\n",
    "    global _dp\n",
    "    for i in range(len(_s2)):\n",
    "        for j in range(len(_s1)):\n",
    "            if i > 0 and j > 0:\n",
    "                _dp[i][j] = max(_dp[i - 1][j - 1] + h(_s2[i], _s1[j]),\n",
    "                                _dp[i - 1][j] + GAP_PENALTY,\n",
    "                                _dp[i][j - 1] + GAP_PENALTY)\n",
    "            elif i > 0:\n",
    "                _dp[i][j] = _dp[i - 1][j] + GAP_PENALTY\n",
    "            elif j > 0:\n",
    "                _dp[i][j] = _dp[i][j - 1] + GAP_PENALTY\n",
    "            else:\n",
    "                _dp[i][j] = 0\n",
    "\n",
    "def print_dp_matrix():\n",
    "    for c in range(len(_s1)):\n",
    "        print(' {0}'.format(_s1[c]).center(4, ' '), end='')\n",
    "    print()\n",
    "    for i in range(len(_s2)):\n",
    "        print('{0}'.format(_s2[i]), end='')\n",
    "        for j in range(len(_s1)):\n",
    "            print('{0}'.format(_dp[i][j]).center(4, ' '), end='')\n",
    "        print()\n",
    "\n",
    "def lbacktrace():\n",
    "    s1_alignment = ''\n",
    "    s2_alignment = ''\n",
    "    i = len(_s2) - 1\n",
    "    j = len(_s1) - 1\n",
    "    while i > 0 or j > 0:\n",
    "        if i > 0 and j > 0:\n",
    "            diag = _dp[i - 1][j - 1]\n",
    "            up = _dp[i - 1][j]\n",
    "            left = _dp[i][j - 1]\n",
    "            if _dp[i][j] == 0:\n",
    "                if len(s2_alignment.strip()) > 0: #done with local alignment, abort\n",
    "                    i = 0\n",
    "                else:\n",
    "                    s1_alignment = _s1[j] + s1_alignment\n",
    "                    s2_alignment = ' ' + s2_alignment\n",
    "                    i -= 1\n",
    "                    j -= 1\n",
    "            elif (_dp[i][j] - diag) == h(_s2[i],_s1[j]):\n",
    "                s1_alignment = _s1[j] + s1_alignment\n",
    "                s2_alignment = _s2[i] + s2_alignment\n",
    "                i -= 1\n",
    "                j -= 1\n",
    "            elif (_dp[i][j] - up) == GAP_PENALTY:\n",
    "                s1_alignment = '*' + s1_alignment\n",
    "                s2_alignment = _s2[i] + s2_alignment\n",
    "                i -= 1\n",
    "            elif (_dp[i][j] - left) == GAP_PENALTY:\n",
    "                s1_alignment = _s1[j] + s1_alignment\n",
    "                s2_alignment = '*' + s2_alignment\n",
    "                j -= 1\n",
    "        elif i > 0:  # up\n",
    "            s1_alignment = '*' + s1_alignment\n",
    "            s2_alignment = _s2[i] + s2_alignment\n",
    "            i -= 1\n",
    "        elif j > 0:  # left\n",
    "            s1_alignment = _s1[j] + s1_alignment\n",
    "            s2_alignment = ' ' + s2_alignment\n",
    "            j -= 1\n",
    "    return s1_alignment, s2_alignment\n",
    "\n",
    "def backtrace():\n",
    "    s1_alignment = ''\n",
    "    s2_alignment = ''\n",
    "    i = len(_s2) - 1\n",
    "    j = len(_s1) - 1\n",
    "    while i > 0 or j > 0:\n",
    "        if i > 0 and j > 0:\n",
    "            diag = _dp[i - 1][j - 1]\n",
    "            up = _dp[i - 1][j]\n",
    "            left = _dp[i][j - 1]\n",
    "            if (_dp[i][j] - up) == GAP_PENALTY:\n",
    "                s1_alignment = '*' + s1_alignment\n",
    "                s2_alignment = _s2[i] + s2_alignment\n",
    "                i -= 1\n",
    "            elif (_dp[i][j] - diag) == h(_s2[i],_s1[j]):\n",
    "                s1_alignment = _s1[j] + s1_alignment\n",
    "                s2_alignment = _s2[i] + s2_alignment\n",
    "                i -= 1\n",
    "                j -= 1\n",
    "            elif (_dp[i][j] - left) == GAP_PENALTY:\n",
    "                s1_alignment = _s1[j] + s1_alignment\n",
    "                s2_alignment = '*' + s2_alignment\n",
    "                j -= 1\n",
    "        elif i > 0:  # up\n",
    "            s1_alignment = '*' + s1_alignment\n",
    "            s2_alignment = _s2[i] + s2_alignment\n",
    "            i -= 1\n",
    "        elif j > 0:  # left\n",
    "            s1_alignment = _s1[j] + s1_alignment\n",
    "            s2_alignment = '*' + s2_alignment\n",
    "            j -= 1\n",
    "    return s1_alignment, s2_alignment\n",
    "\n",
    "def lalign(s1,s2):\n",
    "    set_globs(s1.upper(),s2.upper())\n",
    "    print('DP matrix:')\n",
    "    fill_l_dp_matrix()\n",
    "    print_dp_matrix()\n",
    "    return lbacktrace()\n",
    "\n",
    "def galign(s1,s2):\n",
    "    set_globs(s1.upper(),s2.upper())\n",
    "    print('DP matrix:')\n",
    "    fill_dp_matrix()\n",
    "    print_dp_matrix()\n",
    "    return backtrace()\n",
    "\n",
    "r = 'INHBRADLCD'\n",
    "l = 'IhHpRwDiCp'\n",
    "\n",
    "lalign1,lalign2 = lalign(r,l)\n",
    "print(lalign1+'\\n'+lalign2)\n",
    "ma = MultipleAlignment()\n",
    "ma.add_sequence(lalign1.upper())\n",
    "ma.add_sequence(lalign2.upper())\n",
    "print('Sum of Pairs = {0}'.format(\n",
    "        ma.sum_aligned_char_scores(sum_of_pairs, h)))\n",
    "\n",
    "galign1,galign2 = galign(r,l)\n",
    "print(galign1+'\\n'+galign2)\n",
    "gma = MultipleAlignment()\n",
    "gma.add_sequence(galign1.upper())\n",
    "gma.add_sequence(galign2.upper())\n",
    "print('Sum of Pairs = {0}'.format(\n",
    "        gma.sum_aligned_char_scores(sum_of_pairs, h)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### b. 4) Discuss the alignment."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5: Describe a scenario where a researcher would be interested in investigating both the local and global alignment of two sequences."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 6: Determine if two DNA strings are circular rotations of each other. For example TTGATC is a circular rotation of ATCTTG. You must state all assumptions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n"
     ]
    }
   ],
   "source": [
    "# This code assumes D1 and D2 can match exactly (no \n",
    "# indels, different lengths, mismatches, etc).\n",
    "\n",
    "D1 = 'TTGATC'\n",
    "D2 = 'ATCTTG'\n",
    "\n",
    "\n",
    "def circular_suffix_array(dna_string):\n",
    "    \"\"\"\n",
    "    No need to add a terminal character as circles don't\n",
    "    end.\n",
    "    :param dna_string: a dna string\n",
    "    :return: a circular suffix array of dna_string\n",
    "    \"\"\"\n",
    "    csa = []\n",
    "    [csa.append(dna_string[substring:] + dna_string[:substring]) for substring in range(len(dna_string))]\n",
    "    return sorted(csa)\n",
    "\n",
    "def bwt(dna_string):\n",
    "    \"\"\"\n",
    "    Just ripping the last characters off the circular\n",
    "    suffix array.\n",
    "    :param dna_string: a dna string\n",
    "    :return: a BWT string representation\n",
    "    \"\"\"\n",
    "    csa = circular_suffix_array(dna_string)\n",
    "    return ''.join([suffix[-1:] for suffix in csa])\n",
    "\n",
    "def rotations_of_each_other(d1,d2):\n",
    "    \"\"\"\n",
    "    Checking for matching BWT's\n",
    "    :param d1: a dna string to compare\n",
    "    :param d2: a dna string to compare\n",
    "    :return: True if d1,d2 are rotations, else False\n",
    "    \"\"\"\n",
    "    return bwt(d1) == bwt(d2)\n",
    "\n",
    "print(rotations_of_each_other(D1,D2))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 7: Given a suffix tree of S, count the number of distinct substrings of S. This should run in O(n)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "16\n"
     ]
    }
   ],
   "source": [
    "#representing the suffix tree for AATATT as a list of nested tuples\n",
    "S = [\n",
    "    ('T',\n",
    "     (('','ATT$'),\n",
    "      ('','T$'),\n",
    "      ('','$'))\n",
    "    ),\n",
    "    ('A',\n",
    "     (('','ATATT$'),\n",
    "     ('T',\n",
    "      (('','ATT$'),\n",
    "      ('','T$'))))\n",
    "    ),\n",
    "    ('','$')]\n",
    "\n",
    "L = 0 # label\n",
    "D = 1 # data\n",
    "\n",
    "def terminal(node):\n",
    "    \"\"\"\n",
    "    # is this a terminal node?\n",
    "    :param node: a node in a suffix tree\n",
    "    :return: True if node is terminal, else False\n",
    "    \"\"\"\n",
    "    return node[L] == ''\n",
    "\n",
    "def distinct_substr_counter(st):\n",
    "    \"\"\"\n",
    "    Classic BFS w inner node & terminal node length counting.\n",
    "    By summing the lengths of the inner nodes of our tree and\n",
    "    our terminal (leaf) nodes, we are able to count the distinct\n",
    "    substrings of our suffix tree. This leverages the non-redundant\n",
    "    structure of the tree and is able to achieve a runtime of O(n).\n",
    "    \"\"\"\n",
    "    count = 0\n",
    "    for node in st: # bound by 2 x |S| (Bioinformatics Algorithms Volume 2, p. 131)\n",
    "        if terminal(node):\n",
    "            count += len(node[D]) - 1\n",
    "        else:\n",
    "            count += len(node[L])\n",
    "            \"\"\"\n",
    "            I'm unsure how Python handles this but it can be performed in\n",
    "            constant time in a language that allows direct memory access (C/C++).\n",
    "            We just need to append a reference to this node to the list we're\n",
    "            iterating over.\n",
    "            \"\"\"\n",
    "            [st.append(x) for x in node[D]]\n",
    "    return count\n",
    "\n",
    "print(distinct_substr_counter(S))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.4.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
